Improve clip detection

krisvanneste opened the pull request:
In this branch, I try to improve clip detection by checking for the presence of peaks in the first and last few bins of the amplitude histogram instead of only the presence of more than 1 peak to identify clipping.

I made the following changes:
- decreased smoothing for Gaussian kernel density
- do  not estimate kernel density beyond the max. amplitude
- add 1 bin to kernel density at start and end before peak finding, and remove it again afterwards
- only consider the presence of peaks in the first or last few (5) bins to indicate clipping
- require a minimum of 11 bins in the plotted histogram, and fixed its shift with respect to the kernel density

See also issue https://github.com/SeismicSource/sourcespec/issues/23

claudiodsf commented:
Hi Kris,
thanks for this proposal. 

I tested your code and it works well for "classical" clipped traces (i.e., many samples close to maximum and minimum amplitude), but not with my seismograms, which exhibit strange patterns.

I thought about it and came up with the idea of defining a "clipping score" based on the shape of the kernel density.

Could you please take a look at #25 and tell me what you think?

Thanks!

claudiodsf commented:
Here's an example for which you approach does not work (it works when changing the number of first/last bins to check for peaks).

[WI.ABD.mseed.gz](https://github.com/SeismicSource/sourcespec/files/10813145/WI.ABD.mseed.gz)

![WI ABD 00 HHE](https://user-images.githubusercontent.com/1390993/220897280-f4e3d04c-dae0-4bc3-936b-458a5de3d76a.png)

krisvanneste commented:
For the example you show here, did you apply the clipping detection to the full trace or to the S window?

claudiodsf commented:
Hi again, do you mind checking your approach with this example (from which the above trace comes)?

[sspec_test-20230120-Guadeloupe.tgz](https://github.com/SeismicSource/sourcespec/files/10840119/sspec_test-20230120-Guadeloupe.tgz)

claudiodsf commented:
A few suggestions for improvement:

- making it possible to select the number of first/last bins for checking peaks (user configurable parameter)
- check whether the `sensitivity` parameter (related to peak prominence) is still relevant, or if `min_prominence` can be hardcoded , leaving the above "bins" parameter as the only user configurable parameter
- print some quality metrics (i.e., number of peaks and bin within which the peak is) for any trace (clipped and unclipped), so that the user has a better clue on how to set up the detector (that was the idea behind my "clipping score" which is always printed).

Thanks!

krisvanneste commented:
Claudio, thanks.

I will try to implement your suggestions and test the code on your example. Could you also give me the P- and S-wave arrival times for that? I don't think they are included in the quakeml file.

claudiodsf commented:
> Could you also give me the P- and S-wave arrival times for that? I don't think they are included in the quakeml file.

I don't have phase picks for this example:  I let the code compute theoretical picks.
Is that a problem in your workflow?

krisvanneste commented:
For testing the clipping detection, I directly use the functions in clipping_detection.py without running sourcespec and trim the P- or S-window before. However, I can also estimate them from the seismogram.

claudiodsf commented:
> However, I can also estimate them from the seismogram

Thanks 🙏

krisvanneste commented:
Claudio,

I have tried to implement your suggestions and pushed the changes to my improve_clip_detection branch on github.
The main changes are:
- Added `lower_clip_bound` argument (default value 90) that translates into the number of edge bins in which we look for a peak. It corresponds to the lower bound of the amplitude range (expressed as percentage) to consider as potentially clipped. If you can find a better name for this parameter, let me know.
- Based on my analysis of peak prominences of clipped/unclipped records in my examples, I changed the `min_prominences` corresponding to the different sensitivities.
- Compute densities from min. to max. amplitude instead of +/- max. abs. amplitude, as I found an asymmetry in many cases (even though demeaning is applied to the traces).
- Print information (index, prominence) of detected peaks when debugging is on.
- Indicate amplitude range corresponding to `lower_clip_bound` in debug plot.

With my examples, I obtain the following results for the different sensitivities:
- Sensitivity 1: 25 correct, 10 false negative, 0 false positive
- Sensitivity 2: 31 correct, 4 false negative, 0 false positive
- Sensitivity 3: 31 correct, 4 false negative, 0 false positive
- Sensitivity 4: 33 correct, 1 false negative, 0 false positive
- Sensitivity 5: 34 correct, 0 false negative, 14 false positive

I tried to find a value (in between sensitivity 4 and 5) that would give me 0 false negatives and 0 false positives, but I failed. Based on the testing, I also found out that 1 record marked as clipped in our database was actually not clipped, and another one that was marked as unclipped was actually clipped, so the detection is not bad! Based on these results, I think sensitivities 1 and 2 can be removed, keeping only 3 sensitivity levels. I'm not sure it would be a good idea to settle for 1 fixed value, at least not until many more (un)clipped records have been tested (but I don't have much time for that).

For your Guadeloupe record, the `lower_clip_bound` value has to be slightly decreased (to 88) to detect the clipping. Using only the S phase or P + S (with `signal_pre_time=1` and `win_length=30` as in your configuration file), clipping is already detected at a sensitivity of 1. Using the full record, detection occurs at a sensitivity of 4. The plots below illustrate the 3 cases:

![Guadeloupe_S](https://user-images.githubusercontent.com/2128835/222128207-41ddd36a-e972-4eb5-ab57-34856dd46fee.png)

![Guadeloupe_P+S](https://user-images.githubusercontent.com/2128835/222128220-6c522973-c18d-4730-9423-b5a8b33fc5e1.png)

![Guadeloupe_full](https://user-images.githubusercontent.com/2128835/222128230-776da9ea-f791-4ec0-bb2e-db7839a6b8ec.png)

If I apply the same value for `lower_clip_bound` to my examples, I get 2 false positives with a sensitivity of 4, which is not so bad. So maybe we can consider a lower default value for this parameter than the present one (90).

Let me know what you think.

claudiodsf commented:
Thanks Kris! 

I'll make some test on my side and let you know.

I will also add the relevant config parameters.

As soon as you can give me a feedback on the new approach on #25 (no rush!), I will try and merge the two branches to have the two approaches included in the code (and user-selectable).

krisvanneste commented:
Claudio,
OK, I will change to your branch now and report back later.

claudiodsf commented:
> * If you can find a better name for this parameter, let me know.

What do you think of `clipping_percentile` (see my last commit)?

Note that `lower_clip_bound = 100 - clipping_percentile`

claudiodsf commented:
This PR is now superseded by #25.

Closing here and continuing development there!
